frontend/controller/actions/identity.js

```diff
@@ -370,23 +361,53 @@ export default (sbp('sbp/selectors/register', {
       : []
 
     try {
-      sbp('appLogs/startCapture', username)
-      const state = await sbp('gi.db/settings/load', username)
-      let contractIDs = []
+      sbp('appLogs/startCapture', identityContractID)
+      const { encryptionParams, value: state } = await sbp('gi.db/settings/loadEncrypted', identityContractID, password && ((stateEncryptionKeyId, salt) => {
+        return deriveKeyFromPassword(CURVE25519XSALSA20POLY1305, password, salt + stateEncryptionKeyId)
+      }))
+
+      const contractIDs = []
+      const groupsToRejoin = []
       // login can be called when no settings are saved (e.g. from Signup.vue)
       if (state) {
         // The retrieved local data might need to be completed in case it was originally saved
         // under an older version of the app where fewer/other Vuex modules were implemented.
         sbp('state/vuex/postUpgradeVerification', state)
         sbp('state/vuex/replace', state)
         sbp('chelonia/pubsub/update') // resubscribe to contracts since we replaced the state
-        contractIDs = Object.keys(state.contracts)
+        contractIDs.push(...Object.keys(state.contracts))
+
+        // Some of the group contracts in the state may not have completed the
+        // two-step join process (i.e., an OP_KEY_SHARE had not been received
+        // after joining the last time the state was saved)
+        // The following identifies those groups and saves them to groupsToRejoin
+        groupsToRejoin.push(...contractIDs.filter((contractID) => {
+          return (
+            // (1) We're looking for group contracts
+            state.contracts[contractID].type === 'gi.contracts/group' &&
+            // (2) That potentially haven't been joined by us
+            //     (in which case state.profiles?.[username] will be undefined)
+            !state.profiles?.[username]
+          )
+        }))
       }
       if (!contractIDs.includes(identityContractID)) {
         contractIDs.push(identityContractID)
       }
-      await sbp('gi.db/settings/save', SETTING_CURRENT_USER, username)
-      sbp('state/vuex/commit', 'login', { username, identityContractID })
+      await sbp('gi.db/settings/save', SETTING_CURRENT_USER, identityContractID)
+
+      const loginAttributes: Object = { identityContractID, encryptionParams }
+
+      // If username was not provided, retrieve it from the state
+      // TODO: This is temporary until username is no longer needed internally
+      // in contracts
+      if (username) {
+        loginAttributes.username = username
+      } else {
+        loginAttributes.username = Object.entries(state.namespaceLookups).find(([k, v]) => v === identityContractID)?.[0]
+      }
```

Comment:
This can be simplified:

```js
const loginAttributes = { identityContractID, encryptionParams, username }
if (!loginAttributes.username) {
  loginAttributes.username = Object.entries(state.namespaceLookups).find(([k, v]) => v === identityContractID)?.[0]
}
```